plugins/parsers/bindata/parser.go

```diff
@@ -0,0 +1,216 @@
+	switch timeFormat {
+	case "":
+		timeFormat = defaultTimeFormat
+	case "unix", "unix_ms", "unix_us", "unix_ns":
+	default:
+		return nil, fmt.Errorf("invalid time format %s", timeFormat)
+	}
+
+	// Endiannes
+	var byteOrder binary.ByteOrder
+	endiannes = strings.ToLower(endiannes)
+	switch endiannes {
+	case "", "be":
+		byteOrder = binary.BigEndian
+	case "le":
+		byteOrder = binary.LittleEndian
+	default:
+		return nil, fmt.Errorf("invalid endiannes %q", endiannes)
+	}
+
+	// String encoding
+	if stringEncoding == "" {
+		stringEncoding = defaultStringEncoding
+	}
+	stringEncoding = strings.ToUpper(stringEncoding)
+	if stringEncoding != defaultStringEncoding {
+		return nil, fmt.Errorf(`invalid string encoding %s`, stringEncoding)
+	}
+
+	// Field types, names and sizes
+	knownFields := make(map[string]bool)
+	for i, field := range fields {
+		fieldType, ok := fieldTypes[strings.ToLower(field.Type)]
+		if !ok {
+			return nil, fmt.Errorf(`invalid field type %s`, fields[i].Type)
+		}
+
+		if field.Type == "padding" {
+			// Ignore padding fields
+			continue
+		}
+
+		// Check for duplicate field names
+		fieldName := field.Name
+		if _, ok := knownFields[fieldName]; ok {
+			return nil, fmt.Errorf(`duplicate field name %s`, fieldName)
+		}
+		knownFields[fieldName] = true
+
+		// Time field type check
+		if fieldName == "time" {
+			switch timeFormat {
+			case "unix":
+				if field.Type != "int32" {
+					return nil, fmt.Errorf(`invalid time type, must be int32`)
+				}
+			case "unix_ms", "unix_us", "unix_ns":
+				if field.Type != "int64" {
+					return nil, fmt.Errorf(`invalid time type, must be int64`)
+				}
+			}
+		}
+
+		// Overwrite non-string and non-padding field size
+		if field.Type != "string" {
+			fields[i].Size = uint(fieldType.Size())
+		}
+	}
+
+	return &BinData{
+		metricName:     metricName,
+		timeFormat:     timeFormat,
+		endiannes:      endiannes,
+		byteOrder:      byteOrder,
+		stringEncoding: stringEncoding,
+		fields:         fields,
+		DefaultTags:    defaultTags,
+	}, nil
+}
+
+// SetDefaultTags implements Parser.SetDefaultTags()
+func (binData *BinData) SetDefaultTags(tags map[string]string) {
+	binData.DefaultTags = tags
+}
+
+// Parse implements Parser.Parse()
+func (binData *BinData) Parse(data []byte) ([]telegraf.Metric, error) {
+
+	fields := make(map[string]interface{})
+	var offset uint = 0
+	for _, field := range binData.fields {
+		if offset > uint(len(data)) || offset+field.Size > uint(len(data)) {
+			return nil, fmt.Errorf("invalid offset/size in field %s", field.Name)
+		}
+		if field.Type != "padding" {
+			fieldBuffer := data[offset : offset+field.Size]
+			switch field.Type {
+			case "string":
+				fields[field.Name] = string(fieldBuffer)
+			default:
+				fieldValue := reflect.New(fieldTypes[field.Type])
+				byteReader := bytes.NewReader(fieldBuffer)
+				binary.Read(byteReader, binData.byteOrder, fieldValue.Interface())
+				fields[field.Name] = fieldValue.Elem().Interface()
+			}
+		}
```

Comment:
To be honest I'd like to see something like
```suggestion
		switch field.Type {
		case "padding":
			continue
		case "bool":
			var v bool
			r := bytes.NewReader(data[offset : offset+1])
			if err := binary.Read(r, binData.byteOrder, &v); err != nil {
				return nil, err
			}
			fields[field.Name] = v
		case "uint8":
			var v uint8
			r := bytes.NewReader(data[offset : offset+1])
			if err := binary.Read(r, binData.byteOrder, &v); err != nil {
				return nil, err
			}
			fields[field.Name] = v
		case "int8":
			...
		case "uint16":
			...
		case "int16":
			...
		case "uint32":
			...
		case "int32":
			...
		case "uint64":
			...
		case "int64":
			...
		case "float32":
			...
		case "float64":
			...
		case "string":
			fields[field.Name] = string(data[offset:offset+field.Size)
		}
```

Reply:
Totally agree with padding in switch. Regarding reflection - I though it was cool :) is reflection against IndluxData style/rules?

Reply:
Well it's probably a matter of taste, but I think the switch/case above is much more readable/understandable compared to reflection. There also might be a performance impact, but that's not my primary concern... So please switch to the switch-statement. :-)

Reply:
Regarding the null-terminated strings. How about, if a length is given, we respect this length, otherwise we go for null-termination. This would allow to read non-null-terminated strings (i.e. fixed length fields) which you sometimes see in embedded devices.

Reply:
null-terminated strings - sounds good, good point